### PR TITLE
Feature #12 debian

### LINKED ISSUE #12: Update METbaseimage to complete the transition to the Debian base image
## Describe the New Feature ##
dtcenter/MET#2669 adds a new dependency on the Proj library (recommend using the latest Proj version 9.2.1).

This issue is to update the METbaseimage to include the proj libraries to facilitate the compilation of MET for GitHub actions. This can be accomplished as follows:
1. Update METbaseimage to fully switch over to the Debian base image. That version is currently available on DockerHub as `dtcenter/met-base:v2.0_debian`. And we've been using that base image for MET development for > 6 months.
2. Complete the transition by
  - Replacing `Dockerfile` with the contents of `Dockerfile.debian`
  - Replace `Dockerfile.debian_unit_test_env` with the contents of `Dockerfile.unit_test_env`
  - Updating the file names used in the GHA job
3. Make sure that MET's `compile_MET_all.sh` script and corresponding tar file has been updated to support the compilation of Proj. Note that dtcenter/METbaseimage#13 will soon add support for ecKit and Atlas.
4. Tag this repository with version 2.1 which will push a new METbaseimage version.

Including this issue on the `MET-12.0.0 Development` project since this change relates directly to the compilation of MET. 

### Acceptance Testing ###
Verify that the development version of MET can compile with these changes.

### Time Estimate ###
4 hours.

### Sub-Issues ###
Consider breaking the new feature down into sub-issues.

### Relevant Deadlines ###
Please do as early as possible in the MET version 12.0.0 development timeline.

### Funding Source ###
*Define the source of funding and account keys here or state NONE.*

### Assignee ###
- [x] Select **engineer(s)** or **no engineer** required
- [ ] Select **scientist(s)** or **no scientist** required

### Labels ###
- [x] Select **component(s)**
- [x] Select **priority**
- [x] Select **requestor(s)**

### Projects and Milestone ###
- [x] Select **Repository** and/or **Organization** level **Project(s)** or add **alert: NEED CYCLE ASSIGNMENT** label
- [x] Select **Milestone** as the next official version or **Future Versions**

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [x] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdataio](https://github.com/dtcenter/METdataio/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)

## New Feature Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding source**.
- [ ] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Development** issues
Select: **Repository** level development cycle **Project** for the next official release
Select: **Milestone** as the next official version
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

### PR DESCRIPTION
**Pull request by JohnHalleyGotway**
## Pull Request Testing ##

Note that coordinating changes between this repo and MET is a little difficult. This PR is pulling the compile script from MET’s develop branch even though the develop branch doesn’t contain the updates yet. They haven’t been merged in from the feature_2669_proj branch yet.

- [x] Describe testing already performed for these changes:</br>
Created v2.1_beta4 tag and see that this GHA run successfully built the base image and pushed the result to DockerHub:
https://github.com/dtcenter/METbaseimage/actions/runs/6005438417

Note that this GHA proves that MET can be compiled in this Docker environment:
https://github.com/dtcenter/MET/actions/runs/6005888920

- [x] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>
Review the changes.
Confirm the full transition to Debian base image and elimination of centos.
Note the addition of the proj library.

- [x] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[Yes]**

- [x] Do these changes include sufficient testing updates? **[Yes]**

- [x] Will this PR result in changes to the test suite? **[No]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>

- [x] Please complete this pull request review by **[8/29/23]**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Review the source issue metadata (required labels, projects, and milestone).
- [ ] Complete the PR definition above.
- [ ] Ensure the PR title matches the feature or bugfix branch name.
- [ ] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)**
Select: **Organization** level software support **Project** or **Repository** level development cycle **Project**
Select: **Milestone** as the version that will include these changes
- [ ] After submitting the PR, select the :gear: icon in the **Development** section of the right hand sidebar. Search for the issue that this PR will close and select it, if it is not already selected.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

**Review by georgemccabe — CHANGES_REQUESTED** [review]
I was going to suggest that the unit tests be run to ensure that the changes to the base image work as expected. It looks like the PR run for dtcenter/MET#2672 had a couple seg faults, so that should be resolved before this is approved. I am going to click "Request changes" but if no changes are required to fix the issues in the MET unit tests, then I can approve, because the changes look good as far as I can tell.

Some additional comments:

* I see that the following tools are now built using `apt install` to add the proj library:
  * cmake
  * libtiff-dev
  * sqlite3
  * libsqlite3-dev

The GitHub diff is a bit hard to follow since the Dockerfiles were renamed, so I wanted to note the list of new tools here so it is clear if the changes in this PR are revisited later.

* I see that the deprecated set-output command was replaced with [devops-actions/action-get-tag](https://github.com/marketplace/actions/get-current-git-tag). The command could be replaced with the `echo "{name}={value}" >> $GITHUB_OUTPUT` syntax. This action was created by a verified creator, so it seems like a good solution that can be more easily updated if anything changes on GitHub's end (e.g. update this action to a later version instead of determining the new correct syntax to extract the tag name).

**Comment by JohnHalleyGotway** [comment]
I'd like to proceed with merging this PR and creating a new v2.1 tagged release. I'll re-request this review.

File: Dockerfile
Comment (georgemccabe) [inline Dockerfile:58]: We will need to remember to change this path to no longer reference the feature branch once dtcenter/MET#2672 is merged.

File: README.md
Comment (georgemccabe) [inline README.md:14]:
```suggestion
* Completes transition to debian base image using Debian 12 (bookworm)
```

**Comment by JohnHalleyGotway** [comment]
Note, this [GHA run](https://github.com/dtcenter/MET/actions/runs/6054105774) for the dtcenter/MET#2672 PR runs successfully using the dtcenter/met-base:v2.1 Docker image. Differences are flagged in the output as described in [this comment](https://github.com/dtcenter/MET/pull/2672#issuecomment-1703675581). We discussed these diffs during the METplus Engineering meeting on 9/5/23 and decided to proceed with these changes.

The next steps are:
1. Approve/merge this PR.
2. Delete the existing v2.1 images from DockerHub.
3. Create a new v2.1 release for METbaseimage, along with release notes.
4. Let's GHA build/push the v2.1 image back up to DockerHub.
5. Approve/merge the dtcenter/MET#2672 PR.
6. Create a METplus issue to update the truth data to move past the changes that will likely be introduced by upgrading to Debian 12 with a newer GCC version.

**Review by georgemccabe — APPROVED** [review]
I made a few in-line notes:
* Make sure to update the URL in the Dockerfile after the MET PR is merged because the branch listed will no longer exist
* Mention update from Debian 10 to 12 in the v2.1 release notes

I see that the MET PR tests are still running, but the diffs were explained in [this comment](https://github.com/dtcenter/MET/pull/2672#issuecomment-1703675581). We discussed this in the METplus Engineering meeting on 9/5/2023 and decided that we are OK with these diffs.

I approve as long as there are no explainable diffs that pop up besides the ones that were described above and the minor changes are made at the appropriate time.

**Comment by georgemccabe** [comment]
A note about the order of merging of this and the corresponding MET PR:

To test the changes for dtcenter/MET#2672, we had to create the `dtcenter/met-base:v2.1` DockerHub tag that uses the build script from the MET `feature_2669_proj` branch. We will update the Dockerfile in this PR to no longer reference the feature branch, as it will go away once the MET PR is merged.
We discussed that we should merge dtcenter/MET#2672 pull request, then update this branch to use `develop` instead of `feature_2669_proj` in the URL to the build script. We can retag `v2.1` in METbaseimage after it is merged, which will kick off a rebuild of the Docker image. This last step is not mandatory, but would be good to do to ensure that we don't introduce a typo in the METbaseimage PR that we will have to track down later.

**Pull request merged**
